### PR TITLE
Package specific visibility macros

### PR DESCRIPTION
**Pull request by wjwwood**
Connects to ros2/ros2#37.

Still testing on Windows.

### CODE REVIEW

File: rmw_connext_cpp/CMakeLists.txt [hunk 1]
```diff
@@ -27,7 +27,14 @@ ament_target_dependencies(rmw_connext_cpp
   "rosidl_generator_cpp"
   "rosidl_typesupport_connext_cpp"
   "Connext")
+# On Windows this adds the -DRMW_BUILDING_DLL definition.
+# On Unix (GCC or Clang) it hides the symbols with -fvisibility=hidden.
 configure_rmw_library(rmw_connext_cpp)
+# Additionally, on Windows, add the -DROSIDL_TYPESUPPORT_CONNEXT_CPP_BUILDING_DLL definition.
+if(WIN32)
+  set_target_properties(rmw_connext_cpp PROPERTIES
+    COMPILE_FLAGS "-DROSIDL_TYPESUPPORT_CONNEXT_CPP_BUILDING_DLL")
```
Comment (dirk-thomas) [inline rmw_connext_cpp/CMakeLists.txt]: Why not use `target_compile_definitions` as in the other PR?

File: rmw_connext_cpp/CMakeLists.txt [hunk 2]
```diff
@@ -27,7 +27,14 @@ ament_target_dependencies(rmw_connext_cpp
   "rosidl_generator_cpp"
   "rosidl_typesupport_connext_cpp"
   "Connext")
+# On Windows this adds the -DRMW_BUILDING_DLL definition.
+# On Unix (GCC or Clang) it hides the symbols with -fvisibility=hidden.
```
Comment (dirk-thomas) [inline rmw_connext_cpp/CMakeLists.txt]: Rather then adding this comment to every usage of this function the information should be added to the docblock.

File: rmw_connext_cpp/CMakeLists.txt (hunk 1 above)
Comment (wjwwood) [inline rmw_connext_cpp/CMakeLists.txt]: Yeah I missed that.

**Comment by dirk-thomas** [comment]
Beside the comments +1

File: rmw_connext_cpp/CMakeLists.txt (hunk 2 above)
Comment (wjwwood) [inline rmw_connext_cpp/CMakeLists.txt]: I'd like to do both, otherwise it's harder to figure out where the `-DRMW_BUILDING_DLL` is coming from.

Comment (dirk-thomas) [inline rmw_connext_cpp/CMakeLists.txt]: Both sounds good as well.

**Comment by tfoote** [comment]
lgtm

**Comment by esteve** [comment]
+1

**Comment by wjwwood** [comment]
Ok, I got everything working on Windows now and the latest build is successful:

http://54.183.26.131:8080/job/ros2_batch_ci_windows/207/testReport/

There are some failing tests, but I am looking into those separately as they were there already.

**Pull request merged**